File: src/java/foundgine-providers/src/main/java/com/foundgine/providers/storage/sql/SqlPlan.java
package com.foundgine.providers.storage.sql;

import com.foundgine.core.abstractions.AuthorizationPredicate;
import com.foundgine.core.abstractions.EntityId;
import com.foundgine.core.abstractions.FieldId;
import com.foundgine.core.execution.ProviderPlan;
import com.foundgine.core.semantic.query.SemanticSortDirection;
import com.foundgine.providers.storage.sql.query.SqlParameterBinding;
import java.util.List;

/** Physical SQL representation of a provider-independent execution plan. */
public final class SqlPlan extends ProviderPlan {
    private final String commandText;
    private final List<SqlColumnBinding> columns;
    private final List<SqlParameterBinding> parameters;
    private final SqlPaginationPlan pagination;
    private final List<SqlAuthorizationPredicate> authorization;

    public SqlPlan(String commandText, List<SqlColumnBinding> columns, List<SqlParameterBinding> parameters,
                   SqlPaginationPlan pagination, List<SqlAuthorizationPredicate> authorization) {
        super("sql");
        if (commandText == null) throw new NullPointerException("commandText");
        if (columns == null) throw new NullPointerException("columns");
        this.commandText = commandText;
        this.columns = List.copyOf(columns);
        this.parameters = parameters == null ? List.of() : List.copyOf(parameters);
        this.pagination = pagination;
        this.authorization = authorization == null ? List.of() : List.copyOf(authorization);
    }
    public SqlPlan(String commandText, List<SqlColumnBinding> columns) { this(commandText, columns, null, null, null); }
    public String commandText() { return commandText; }
    public List<SqlColumnBinding> columns() { return columns; }
    public List<SqlParameterBinding> parameters() { return parameters; }
    public List<SqlParameterBinding> effectiveParameters() { return parameters; }
    public SqlPaginationPlan pagination() { return pagination; }
    public List<SqlAuthorizationPredicate> authorization() { return authorization; }
}

File: src/csharp/Foundgine.Providers/Storage/Sql/SqlPlan.cs
using Foundgine.Core.Execution;
using Foundgine.Core.Semantic.Metadata;
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Query;

namespace Foundgine.Providers.Storage.Sql;

/// <summary>
/// Physical SQL representation of a provider-independent execution plan.
/// SQL exists only at this provider boundary.
/// </summary>
public sealed record SqlPlan(
    string CommandText,
    IReadOnlyList<SqlColumnBinding> Columns,
    IReadOnlyList<Foundgine.Providers.Storage.Sql.Query.SqlParameterBinding>? Parameters = null,
    SqlPaginationPlan? Pagination = null,
    IReadOnlyList<SqlAuthorizationPredicate>? Authorization = null) : ProviderPlan("sql")
{
    public IReadOnlyList<Foundgine.Providers.Storage.Sql.Query.SqlParameterBinding> EffectiveParameters =>
        Parameters ?? [];
}

public sealed record SqlColumnBinding(
    string ResultName,
    EntityId EntityId,
    FieldId FieldId,
    string ColumnName,
    int NodeId,
    bool IsCursor = false);

/// <summary>
/// Forward keyset pagination metadata. CursorValues describes the exact
/// semantic ordering tuple used by the SQL seek predicate.
/// </summary>
public sealed record SqlPaginationPlan(
    int First,
    IReadOnlyList<SqlCursorBinding> CursorValues,
    string? After);

public sealed record SqlCursorBinding(
    string ResultName,
    EntityId EntityId,
    FieldId FieldId,
    Type ClrType,
    SemanticSortDirection Direction);

/// <summary>
/// Authorization predicates carried into the SQL provider plan. The predicate
/// remains provider-independent here; SQL lowering can bind its members and
/// context values without retaining an expression tree or delegate.
/// </summary>
public sealed record SqlAuthorizationPredicate(
    int NodeId,
    AuthorizationPredicate Predicate);